core/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java

```diff
@@ -97,7 +102,71 @@ public void testRandomClusterStateUpdates() {
+     * imported as dangling indices.  Normally, the first cluster state update that the node
+     * receives from the new cluster would contain a cluster block that would cause all in-memory
+     * structures to be removed (see {@link IndicesClusterStateService#clusterChanged(ClusterChangedEvent)}),
+     * but in the case where the node joined and was a few cluster state updates behind, it would
+     * not have received the cluster block, in which case we still need to remove the in-memory
+     * structures while ensuring the data remains on disk.  This test executes this particular
+     * scenario.
+     */
+    public void testJoiningNewClusterOnlyRemovesInMemoryIndexStructures() {
+        // we have an IndicesClusterStateService per node in the cluster
+        final Map<DiscoveryNode, IndicesClusterStateService> clusterStateServiceMap = new HashMap<>();
+        ClusterState previousState = randomInitialClusterState(clusterStateServiceMap, RecordingIndicesService::new);
+        ClusterState newClusterState = ClusterState.builder(previousState)
+                                           .metaData(MetaData.builder(previousState.metaData()).clusterUUID(UUIDs.randomBase64UUID()))
+                                           .build();
+
+        // add index to the cluster
+        String name = "index_" + randomAsciiOfLength(8).toLowerCase(Locale.ROOT);
+        CreateIndexRequest request = new CreateIndexRequest(name, Settings.builder()
+                                                                      .put(SETTING_NUMBER_OF_SHARDS, randomIntBetween(1, 3))
+                                                                      .put(SETTING_NUMBER_OF_REPLICAS, 1)
+                                                                      .build());
+        ClusterState state = cluster.createIndex(previousState, request);
+        assertTrue(state.metaData().hasIndex(name));
+
+        // pick a data node to simulate the cluster state change on
+        DiscoveryNode node = null;
+        for (DiscoveryNode discoveryNode : clusterStateServiceMap.keySet()) {
+            if (discoveryNode.isDataNode()) {
+                node = discoveryNode;
+                break;
+            }
+        }
+        assertNotNull(node);
+        IndicesClusterStateService indicesClusterStateService = clusterStateServiceMap.get(node);
+        ClusterState localState = adaptClusterStateToLocalNode(state, node);
+        ClusterState previousLocalState = adaptClusterStateToLocalNode(previousState, node);
+        indicesClusterStateService.clusterChanged(
+            new ClusterChangedEvent("simulated change 1", localState, previousLocalState));
+
+        // check that cluster state has been properly applied to node
+        assertClusterStateMatchesNodeState(localState, indicesClusterStateService);
+
+        indicesClusterStateService = clusterStateServiceMap.get(node);
```

Comment:
indicesClusterStateService should not have changed, so no need to refetch from the map.